### ISSUE TITLE
SSL Certificate reread on HUP signal

### ISSUE THREAD

**Issue by acidtonic** [oatpp/oatpp#483]
On Unix it would be nice if oatpp could reread the configuration files related to SSL certificates such that they could be updated without restarting the server.

Normally what would happen is the following...

1. SSL files get updated.
2. Oatpp server receives HUP signal
3. Oatpp rereads the certificate without refusing any connections or dropping existing connections still using the previous certificate.

Although we should have an HA environment in production, In my opinion a failover shouldn't be needed to change certificates and remain operational.

**Comment by bamkrs** [comment]
Hey @acidtonic 
That is indeed a very good point.
However, this will be a tough one. The main issue is in Oat++'s component system. I see two options here. Introduce "Factory-Components" that run their lambda every time the component is requested or implement the `XYZssl` components as some kind of stateful factory which creates new library-instances on HUP signals. 

However, the easiest, but not cleanest way, would be to introduce a new function to the `XYZssl`'s `ConnectionProvider` which you can pass a new configuration. This function would then handle the creation of new library instances etc.

**Comment by bamkrs** [comment]
I created an PR to show my last proposed approach, its a bit overkill since there is an more easy way.

As pointed by @lganzzzo in https://github.com/oatpp/oatpp-openssl/pull/8#issuecomment-919750055 it's way easier to just create an entire new `ConnectionProvider` with a new setter-function for the `ConnectionProvider` in `Server`.
This will need some manual labor since you cant just use the components anymore and have to hand-build them.

**Comment by bamkrs** [comment]
With https://github.com/oatpp/oatpp/pull/484 the workflow would be something like that:

```c++
std::shared_ptr<oatpp::network::Server> server;

std::shared_ptr<oatpp::network::server::ConnectionProvider> createNewProvider() {
    auto config = oatpp::libressl::Config::createDefaultServerConfigShared(CERT_CRT_PATH, CERT_PEM_PATH /* private key */);
    return oatpp::libressl::server::ConnectionProvider::createShared(config, {"0.0.0.0", 8443, oatpp::network::Address::IP_4});
}

void signal_handler(int sig) {
  if (sig == SIG_HUP) {
    server->setConnectionProvider(createNewProvider());
  }
}

void you_run_method() {
  // components, controllers etc
  // [...]

  /* create server */
  server = oatpp::network::Server::createShared(createNewProvider(), OATPP_GET_COMPONENT(std::shared_ptr<ConnectionHandler>));
  server.run()

}
  
```

**Comment by acidtonic** [comment]
Just adding that I agree with either solution. The code example proposed would work for me.

**Comment by bamkrs** [comment]
Hey @acidtonic we just merged the required API for this snippet. Please treat my example es pseudo-code. It isn't tested. 
If I find some spare time tomorrow, I will update our ssl-examples accordingly (with tested code).

**Comment by lganzzzo** [comment]
@bamkrs , @acidtonic 

Actually, it can be done even simpler (don't know why I didn't think of it earlier):

Just create a proxy connection provider - which will switch providers internally if needed

Proxy provider:

```cpp
class ProxyConnectionProvider : public oatpp::network::ServerConnectionProvider {
private:
  std::shared_ptr<oatpp::network::ServerConnectionProvider> m_currentProvider;
public:

  std::shared_ptr<data::stream::IOStream> get() override {
    return m_currentProvider;
  }

  /* call this method when certificates changed and a new SSL connection provider created */
  void resetProvider(const std::shared_ptr<oatpp::network::ServerConnectionProvider>& newProvider) {
    m_currentProvider = newProvider;
  }

}
```

AppComponent:

```cpp
OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::network::ServerConnectionProvider>, serverConnectionProvider)([this] {

  auto sslProvider = <create-ssl-provider>;
  auto proxyProvider = std::make_shared<ProxyConnectionProvider>();
  proxyProvider->resetProvider(sslProvider);

  return proxyProvider;

}());
```

This way, you can use oatpp components as before. Plus we don't have to introduce additional locking in the `network::Server` as we just did to be able to set a new provider

**Comment by bamkrs** [comment]
@lganzzzo @acidtonic 
See https://github.com/oatpp/example-libressl/pull/5 :)

**Comment by acidtonic** [comment]
Beautiful everyone involved. I will get this tested when I have some spare time. We are busy with a local show for awhile so I won't get to this right away.

Thanks again :)

**Comment by lganzzzo** [comment]
Fixed in `1.3.0` see [ConnectionProviderSwitch](https://github.com/oatpp/oatpp/blob/master/changelog/1.3.0.md#connectionproviderswitch)
